File: src/main/java/com/teamwizardry/wizardry/common/spell/component/EffectInstance.java
package com.teamwizardry.wizardry.common.spell.component;

import java.util.Map;

import net.minecraft.nbt.NbtCompound;

/**
 * Contains data relevant to a single cast event of a {@code ModuleEffect} component.
 * Do not construct, instances are provided for calls to
 * {@link PatternEffect#affectBlock(net.minecraft.world.World, Interactor, Instance)} and
 * {@link PatternEffect#affectEntity(net.minecraft.world.World, Interactor, Instance)}
 * See {@link Instance} for detailed information on the available data.
 * @see Instance
 * @see PatternEffect
 */
public class EffectInstance extends Instance {
    public EffectInstance(Pattern pattern, TargetType targetType, Map<String, Double> attributeValues, double manaCost,
                          double burnoutCost, Interactor caster) {
        super(pattern, targetType, attributeValues, manaCost, burnoutCost, caster);
        if (extraData == null) {
            extraData = new NbtCompound();
        }
        extraData.putString(PATTERN_TYPE, "effect");
    }
}

File: src/main/java/com/teamwizardry/wizardry/common/spell/component/EffectInstance.kt
package com.teamwizardry.wizardry.common.spell.component

/**
 * Contains data relevant to a single cast event of a `ModuleEffect` component.
 * Do not construct, instances are provided for calls to
 * [PatternEffect.affectBlock] and
 * [PatternEffect.affectEntity]
 * See [Instance] for detailed information on the available data.
 * @see Instance
 *
 * @see PatternEffect
 */
class EffectInstance(
    pattern: Pattern?, targetType: TargetType, attributeValues: Map<String?, Double>, manaCost: Double,
    burnoutCost: Double, caster: Interactor
) : Instance(pattern, targetType, attributeValues, manaCost, burnoutCost, caster) {
    init {
        if (extraData == null) {
            extraData = NbtCompound()
        }
        extraData.putString(PATTERN_TYPE, "effect")
    }
}